### PR TITLE
ensure that the status channel is always initialized

### PR DESCRIPTION
**Pull request by dimalinux**
This PR adds a new test case, `TestXMRNotLockedAndETHRefundedAfterAliceRestarts`, that was initially broken before the rest of the PR. The taker's status subscribe channel was not getting the refund status change after restart.

I decided to remove the status channel from both the `OfferExtra` and `swap.Info` struct types. Both of those types are designed for serialization (and deserialization on restart), but status channels are ephemeral to a single run of `swapd`. Instead of trying to keep the status channels synchronized across two different data structures, I made management of the channel part of `swap.Manager`.

### CODE REVIEW

File: common/types/offer.go [hunk 1]
```diff
@@ -228,3 +222,19 @@ func (o *Offer) UnmarshalJSON(data []byte) error {
...
+	// to claim.
+	UseRelayer bool `json:"useRelayer,omitempty"`
+}
+
+// NewOfferExtra creates an OfferExtra instance
+func NewOfferExtra(forceUseRelayer bool) *OfferExtra {
+	return &OfferExtra{
+		UseRelayer: forceUseRelayer,
+	}
+}
```
Comment (dimalinux) [inline common/types/offer.go:240]: I was doing custom marshaling to ensure that the `StatusCh` was always initialized in an early version of this PR. Otherwise I wouldn't have made so many edits here. Everyone is calling `NewOfferExtra` now instead of creating the structure directly.

File: daemon/test_support.go [hunk 2]
```diff
@@ -87,13 +88,15 @@ func CreateTestBootnode(t *testing.T) (uint16, string) {
...
+	dataDir := t.TempDir()
+
 	conf := &bootnode.Config{
 		Env:           common.Development,
 		DataDir:       t.TempDir(),
 		Bootnodes:     nil,
 		P2PListenIP:   "127.0.0.1",
 		Libp2pPort:    0,
-		Libp2pKeyFile: common.DefaultLibp2pKeyFileName,
+		Libp2pKeyFile: path.Join(dataDir, common.DefaultLibp2pKeyFileName),
```
Comment (dimalinux) [inline daemon/test_support.go:99]: This was a bug in my previous PR. Tests were littering network key files in the current working directory of the test.

File: protocol/backend/backend.go [hunk 3]
```diff
@@ -262,7 +262,7 @@ func (b *backend) ClearXMRDepositAddress(offerID types.Hash) {
 // HasOngoingSwapAsTaker returns nil if we have an ongoing swap with the given peer where
 // we're the xmrtaker, otherwise returns an error.
 func (b *backend) HasOngoingSwapAsTaker(remotePeer peer.ID) error {
-	swaps, err := b.swapManager.GetOngoingSwaps()
+	swaps, err := b.swapManager.GetOngoingSwapsSnapshot()
```
Comment (dimalinux) [inline protocol/backend/backend.go:265]: This was another change that I made when I was still working on a design that kept the status channel inside the `swap.Info` structure, but it seems worth keeping. Methods that return deep copies and are therefore completely thread safe, have the word `Snapshot` on the end.

File: protocol/swap/status_manager.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+	"github.com/athanorlabs/atomic-swap/common/types"
+)
+
+// statusManager provides lookup for the status channels. Status channels are
+// ephemeral between runs of swapd.
+type statusManager struct {
+	mu             sync.Mutex
+	statusChannels map[types.Hash]chan Status
+}
```
Comment (dimalinux) [inline protocol/swap/status_manager.go:14]: I isolated the status channel management into this separate `statusManager` "class" and gave `swap.manager` an "is-a" relationship with it. The base swap manager was already using its mutex across multiple independent maps. Separating this, hopefully, lowers the code complexity.

File: protocol/swap/status_manager.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+	return sm.getStatusChan(offerID)
+}
+
+// DeleteStatusChan deletes any status channel associated with the offer ID.
+func (sm *statusManager) DeleteStatusChan(offerID types.Hash) {
+	sm.mu.Lock()
+	defer sm.mu.Unlock()
+
+	delete(sm.statusChannels, offerID)
+}
```
Comment (dimalinux) [inline protocol/swap/status_manager.go:49]: No one is using this yet and I'm not entirely clear what the lifecycle should be. For `net_makeOfferAndSubscribe`, I think you want it to live across multiple takers if the first one fails. If we can figure out the lifecycle before merging, that might be good, but I don't think anyone is going to run enough swaps without restarting for this to budge resource usage noticeably.

File: protocol/xmrmaker/message_handler.go [hunk 6]
```diff
@@ -72,7 +72,7 @@ func (s *swapState) setNextExpectedEvent(event EventType) error {
 		panic("status corresponding to event cannot be UnknownStatus")
 	}
 
-	s.info.SetStatus(status)
+	s.UpdateStatus(status)
 	err := s.Backend.SwapManager().WriteSwapToDB(s.info)
```
Comment (dimalinux) [inline protocol/xmrmaker/message_handler.go:76]: When do we write the the DB after updating the status and when do we not do it?

File: rpc/net.go [hunk 7]
```diff
@@ -37,12 +38,12 @@ type NetService struct {
 	net        Net
 	xmrtaker   XMRTaker
 	xmrmaker   XMRMaker
-	sm         SwapManager
+	sm         swap.Manager
```
Comment (dimalinux) [inline rpc/net.go:41]: Nothing here. I just removed the type alias.

File: protocol/swap/status_manager.go (hunk 5 above)
Comment (noot) [reply protocol/swap/status_manager.go:49]: I would expect it to be deleted when the swap ends (no matter the status) and if it's retaken, then a new channel would be created

File: protocol/swap/swap_manager.go [hunk 8]
```diff
@@ -140,40 +147,86 @@ func (m *manager) GetPastSwap(id types.Hash) (*Info, error) {
...
-		swaps[i] = sCopy
-		i++
+		offerIDs = append(offerIDs, &s.OfferID)
 	}
+
+	return offerIDs, nil
+}
+
+// GetOngoingSwapsSnapshot returns a copy of all ongoing swaps. If you need to
+// modify the result or read from its status channel, call `GetOngoingSwap` on
```
Comment (noot) [inline protocol/swap/swap_manager.go:204]: `Info` no longer has a status channel?

File: protocol/swap/swap_manager.go [hunk 9]
```diff
@@ -140,40 +147,86 @@ func (m *manager) GetPastSwap(id types.Hash) (*Info, error) {
...
+	s, has := m.ongoing[offerID]
+	if !has {
+		return nil, errNoSwapWithOfferID
+	}
+
+	// We have the read lock above when making this copy, but the swapState
+	// instances don't use that lock when they modify the same Info structure
+	// from a different go process. At some point we should update the swapState
+	// instances to make all changes to their ongoing Info structure via this
+	// manager.
```
Comment (noot) [inline protocol/swap/swap_manager.go:180]: we should open an issue for this; ideally we don't need the snapshot functions and it's only copies that are ever returned, and the other processes modify the swap info (I think status is the only field modified after creation?) via this manager

File: protocol/swap/status_manager.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+// PushNewStatus adds a new status to the offer ID's channel
+func (sm *statusManager) PushNewStatus(offerID types.Hash, status types.Status) {
```
Comment (noot) [inline protocol/swap/status_manager.go:52]: we should remove `info.SetStatus` and make a function on the swapManager that updates the status that also calls this function

File: protocol/xmrtaker/instance.go [hunk 11]
```diff
@@ -64,12 +64,17 @@ func NewInstance(cfg *Config) (*Instance, error) {
...
 	if err != nil {
 		return err
 	}
 
-	for _, s := range swaps {
+	for _, offerID := range ongoingIDs {
+		s, err := inst.backend.SwapManager().GetOngoingSwap(*offerID)
+		if err != nil {
+			return err
+		}
```
Comment (noot) [inline protocol/xmrtaker/instance.go:76]: why is this changed? can you add a function to get all ongoing swaps like before?

File: protocol/swap/swap_manager.go (hunk 8 above)
Comment (dimalinux) [reply protocol/swap/swap_manager.go:204]: Correct. There is some information in the PR description as to why.

File: protocol/xmrtaker/instance.go (hunk 11 above)
Comment (dimalinux) [reply protocol/xmrtaker/instance.go:76]: `GetOngoingSwapsSnapshot` returns all the swaps (as a deep copy) and it is still there. I thought the code was easier to audit if we just got the IDs here and later pull the non-deep-copies on an individual bases when restarting a swap.

File: protocol/swap/swap_manager.go (hunk 9 above)
Comment (dimalinux) [reply protocol/swap/swap_manager.go:180]: > I think status is the only field modified after creation?

`EndTime`, `LastStatusUpdateTime`, `Timeout1` and `Timeout2` are also modified after creation.

File: protocol/swap/status_manager.go (hunk 5 above)
Comment (dimalinux) [reply protocol/swap/status_manager.go:49]: > I would expect it to be deleted when the swap ends (no matter the status) and if it's retaken, then a new channel would be created

It is handled that way now.

File: protocol/swap/swap_manager.go (hunk 9 above)
Comment (dimalinux) [reply protocol/swap/swap_manager.go:180]: > we should open an issue for this

The locking is now in this PR, so we don't need to open an issue. I removed the comment above. The locking in here is exclusively for the map maintained by the manager. The lock for the `Info` structure is inside the struct and used by member functions of the struct.

File: protocol/swap/status_manager.go (hunk 10 above)
Comment (dimalinux) [reply protocol/swap/status_manager.go:52]: See what I implemented and if you like it. I was originally thinking along the lines of what you wrote above, but I decided that this manager should only deal with maintaining the map of the status channel (which has a different lifecycle than the `Info` struct). Locking and changing the `Info` struct is handled by member functions that are invoked by the maker/taker `swapStage` go-processes.

File: protocol/swap/swap_manager.go (hunk 8 above)
Comment (noot) [reply protocol/swap/swap_manager.go:199]: yeah I meant you need to update the comment here

File: protocol/swap/types.go [hunk 12]
```diff
@@ -63,8 +62,18 @@ type Info struct {
...
+	// rwMu handles synchronization when LastStatusUpdateTime, Timeout1,
+	// Timeout2 and EndTime are updated. This Info struct is modified by the
+	// maker or taker's swapState go process as the state of the swap
+	// progresses. The swapState go process does not need synchronization when
+	// reading its own changes, but it needs to grab a write lock when modifying
+	// the the structure. Readers from other go-processes only get copies of
+	// this structure. They exclusively use the DeepCopy method to get their
+	// copy, which grabs the read lock ensuring that they always capture the
+	// up-to-date state of this Info struct.
+	rwMu sync.RWMutex
```
Comment (noot) [inline protocol/swap/types.go:76]: why did you decide to go with a lock in the info struct instead of having the swap manager `Get` functions be read-only (ie. only return copies) and having swap manager implement `Set` functions which modify the info fields for a given swap?

File: protocol/xmrmaker/instance.go [hunk 13]
```diff
@@ -84,12 +84,17 @@ func NewInstance(cfg *Config) (*Instance, error) {
...
 	if err != nil {
 		return err
 	}
 
-	for _, s := range swaps {
+	for _, offerID := range ongoingIDs {
+		s, err := inst.backend.SwapManager().GetOngoingSwap(*offerID)
+		if err != nil {
+			return err
+		}
```
Comment (noot) [inline protocol/xmrmaker/instance.go:96]: why two iterations? why not do `GetOngoingSwapsSnapshot` directly?

File: protocol/xmrmaker/swap_state.go [hunk 14]
```diff
@@ -452,6 +450,11 @@ func (s *swapState) SendKeysMessage() common.Message {
 	}
 }
 
+func (s *swapState) UpdateStatus(status types.Status) {
```
Comment (noot) [inline protocol/xmrmaker/swap_state.go:453]: doesn't need to be exported I don't think

File: protocol/xmrtaker/instance.go (hunk 11 above)
Comment (noot) [reply protocol/xmrtaker/instance.go:76]: I don't think it makes a difference and this results in 2 iterations instead of 1, and we're iterating through all of them either way

Comment (noot) [reply protocol/xmrtaker/instance.go:76]: maybe implement a `GetOngoingSwaps` that returns all the swaps not as a copy?

File: protocol/xmrtaker/swap_state.go [hunk 15]
```diff
@@ -318,6 +317,11 @@ func (s *swapState) SendKeysMessage() common.Message {
 	}
 }
 
+func (s *swapState) UpdateStatus(status types.Status) {
```
Comment (noot) [inline protocol/xmrtaker/swap_state.go:320]: doesn't need to be exported

**Review by noot — COMMENTED** [review]
basically looks good, mostly minor comments, and an overall question about why the setters were left on `Info` instead of `swap.Manager`. I'm not sure if designing it like that would be nicer or worse :/

File: protocol/swap/types.go (hunk 12 above)
Comment (dimalinux) [reply protocol/swap/types.go:76]: The swap manager's existing locks are for the maps inside the manager. They act globally across all `Info` objects in the ongoing or past swap maps. To manage the state the state of an individual swap, we needed a lock that is specific to that individual `Info` object. The swap manager could maintain a map of locks for the individual `Info` structures, but I thought this solution was cleaner.

File: protocol/xmrmaker/message_handler.go (hunk 6 above)
Comment (dimalinux) [reply protocol/xmrmaker/message_handler.go:76]: @noot Did you see this question? It's not critical to the PR. I just wanted to understand it better.

File: protocol/swap/swap_manager.go (hunk 8 above)
Comment (dimalinux) [reply protocol/swap/swap_manager.go]: Now I understand. I removed the part about the status channel from the comment. Thanks.

File: protocol/xmrtaker/instance.go (hunk 11 above)
Comment (dimalinux) [reply protocol/xmrtaker/instance.go:76]: I downgraded the code from getting the full, modifiable object list here, because there is nothing at this high level in the code that needs access to anything but the offer IDs. By getting the modifiable copies lower down on the stack on an individual basis, it shortens the chain-of-custody for the modifiable objects and reduces the API surface area that returns modifiable objects. Auditing synchronous access to objects is hard and I was hoping these changes make the code easier to audit for correct use of locks.

File: protocol/xmrmaker/instance.go (hunk 13 above)
Comment (dimalinux) [reply protocol/xmrmaker/instance.go:96]: See the comment on the maker. If, after reading it, you still want the change, I can make it.

File: protocol/xmrmaker/swap_state.go (hunk 14 above)
Comment (dimalinux) [reply protocol/xmrmaker/swap_state.go]: Fixed!

File: protocol/xmrtaker/swap_state.go (hunk 15 above)
Comment (dimalinux) [reply protocol/xmrtaker/swap_state.go]: Fixed!

**Review by noot — APPROVED** [review]
